services/galley/test/integration/API/MLS.hs

```diff
@@ -670,6 +705,134 @@ testCommitNotReferencingAllProposals = withSystemTempDirectory "mls" $ \tmp -> d
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" (pClients bob)
+
+  do
+    convs <-
+      responseJsonError =<< getConvs (qUnqualified (pUserId bob)) Nothing Nothing
+        <!! const 200 === statusCode
+    liftIO $
+      assertBool
+        "bob is in conversation before he gets removed"
+        (conversation `elem` map cnvQualifiedId (convList convs))
+
+  testSuccessfulRemoveMemberFromConvCommit creator [bob] conversation removalCommit [bob]
+
+  do
+    convs <-
+      responseJsonError =<< getConvs (qUnqualified (pUserId bob)) Nothing Nothing
+        <!! const 200 === statusCode
+    liftIO $
+      assertBool
+        "bob is not longer part of conversation after the commit"
+        (conversation `notElem` map cnvQualifiedId (convList convs))
+
+testRemoveClientsIncomplete :: TestM ()
+testRemoveClientsIncomplete = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  -- remove only first client of bob
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" [(NE.head (pClients bob))]
+
+  err <-
+    responseJsonError
+      =<< postMessage (qUnqualified (pUserId creator)) removalCommit
+        <!! statusCode === const 409
+
+  liftIO $ Wai.label err @?= "mls-client-mismatch"
+
+testUserRemovesThemselvesFromConv :: TestM ()
+testUserRemovesThemselvesFromConv = withSystemTempDirectory "mls" $ \tmp -> do
+  (creator, [bob]) <- withLastPrekeys $ setupParticipants tmp def [(2, LocalUser)]
+
+  -- create a group
+  (groupId, conversation) <- setupGroup tmp CreateConv creator "group"
+
+  -- add clients to it and get welcome message
+  (addCommit, welcome) <-
+    liftIO $
+      setupCommit tmp creator "group" "group" $
+        NonEmpty.tail (pClients creator) <> toList (pClients bob)
+
+  testSuccessfulCommit MessagingSetup {users = [bob], commit = addCommit, ..}
+
+  -- FUTUREWORK: create commit as bob, when the openmls library supports removing own clients
+  (removalCommit, _mbWelcome) <- liftIO $ setupRemoveCommit tmp creator "group" "group" (pClients bob)
+
+  -- bob tries to leave the conversation by removng all its clients
+  err <-
+    responseJsonError
+      =<< postMessage (qUnqualified (pUserId bob)) removalCommit
+        <!! statusCode === const 409
+
+  liftIO $ Wai.label err @?= "mls-self-removal-not-allowed"
+
+testRemoveDeletedClient :: Bool -> TestM ()
```

Comment:
It's a tiny thing and it doesn't change the outcome, but instead of using `Bool`, I'd introduce a simple sum type:

```haskell
data DeleteClientBefore = DeleteClient | DontDeleteClient
```

Reply:
I don't think there isn't much gained from it, because it is only used in one test and the type signature is right above variable binding `deleteClientBefore` which is descriptive enough